rfd/0067-database-access-aws-redis

```diff
@@ -0,0 +1,221 @@
+
+Here is a sample config for specifying the secret store in database service.
+```
+db_service:
+  ...
+
+  secert_store:
+    type: "aws-secrets-manager"
+    key_prefix: "teleport/"
+```
+
+And here is a sample interface for managing secrets using the secret store.
+```
+// SecretValue is the secret value.
+type SecretValue struct {
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret.
+    Value string
+    // Version is the version of the secret value.
+    Version string
+    // Created is the creation time of this version. Can be used to calculate
+    // if a secret value is expired.
+    Created time.Time
+}
+
+// CreateRequest defines the request to create a new secret.
+type CreateRequest struct {
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret to set. A random value is generated if
+    // empty.
+    Value string
+    // RandomValueLength is the length of the random value to generate.
+    RandomValueLength int
+}
+
+// PutValueRequest defines the request to create a new secret version.
+type PutValueRequest struct {
+    // CurrentValue is the current secret value. Key in CurrentValue is used to
+    // identify the secret. Other attributes can be used for test-and-set
+    // operations.
+    CurrentValue *SecretValue
+    // Value is the value of the secret to set. A random value is generated if
+    // empty.
+    Value string
+    // RandomValueLength is the length of the random value to generate.
+    RandomValueLength int
+}
+
+// Secrets defines an interface for managing secrets. A secret consists of a
+// key path and a list of versions which hold copies of current or past secret
+// values.
+type Secrets interface {
+    // Create creates a new secret.
+    Create(req CreateRequest) (*SecretValue, error)
+
+    // Delete deletes the secret for provided path.
+    Delete(key string) error
+
+    // PutValue creates a new secret version for the secret. Implementations
+    // must guarantee that only the first of many calls to PutValue succeeds,
+    // when req.CurrentValue are the same.
+    PutValue(req PutValueRequest) (*SecretValue, error)
+
+    // GetValue returns the secret value for provided version. Besides version
+    // string returned from PutValue, two specials versions "CURRENT" and
+    // "PREVIOUS" can also be used to retrieve the current and previous
+    // versions respectively. If version is empty, "CURRENT" is used.
+    GetValue(key string, version string) (*SecretValue, error)
+}
+```
+It is up to the caller to call `PutValue` periodically to rotate the secret.
+
+#### Secret store - Random password generation
+[sethvargo/go-password](https://github.com/sethvargo/go-password) can be used to generate random passwords. <sup>[3]<sup>
+
+A minimum length of 16 should be enforced for generating random passwords. A mix of lower case letters, upper case letters, digits, and symobols should be also enforced.
```

Comment:
I would focus on entropy over what character classes to mandate. Now, the keyspace is important, but every password doesn't have to have all these character types. In fact, I personally try to avoid symbols in service-related passwords, as they have a habit of tripping up tools when parsing. We use should the maximum password length available.

Reply:
Removed this section. Added in the security section below to use the maximum password length available. Thanks!